app/models/group_order_invoice.rb

```diff
@@ -0,0 +1,60 @@
+class GroupOrderInvoice < ApplicationRecord
+
+  belongs_to :group_order
+  validates_presence_of :group_order
+  validates_uniqueness_of :invoice_number
+  validate :tax_number_set
+  after_initialize :init, unless: :persisted?
+
+  def generate_invoice_number(count)
+    trailing_number = count.to_s.rjust(4, '0')
+    unless  GroupOrderInvoice.find_by(invoice_number: Time.now.strftime("%Y%m%d") + trailing_number)
+      Time.now.strftime("%Y%m%d") + trailing_number
+    else
+      generate_invoice_number(count.to_i + 1)
```

Comment:
* recursive calling `generate_invoice_number` here is a very bad idea, since you create a database query all the time which is slow and blocks everything else
* you could either store the invoice number somewhere in the database or find the last invoice via a single query to the database (something like `SELECT invoice_number ... WHER invoice_number LIKE '...%' ORDER BY invoice_number DESC LIMIT 1`)